src/main/java/com/box/sdk/RequestBodyFromStream.java

```diff
@@ -13,41 +15,70 @@ final class RequestBodyFromStream extends RequestBody {
     private final InputStream inputStream;
     private final ProgressListener progressListener;
     private final MediaType mediaType;
-    private final int contentLength;
+    private final long fileSize;
 
-    RequestBodyFromStream(InputStream inputStream, MediaType mediaType, ProgressListener progressListener) {
+    RequestBodyFromStream(
+        InputStream inputStream, MediaType mediaType, ProgressListener progressListener, long fileSize
+    ) {
         this.inputStream = inputStream;
         this.progressListener = progressListener;
         this.mediaType = mediaType;
-        try {
-            this.contentLength = inputStream.available();
-        } catch (IOException e) {
-            throw new RuntimeException("Cannot read input stream for upload", e);
-        }
+        this.fileSize = fileSize;
     }
 
     @Override
-    public long contentLength() {
-        return contentLength;
+    public MediaType contentType() {
+        return mediaType;
     }
 
     @Override
-    public MediaType contentType() {
-        return mediaType;
+    public void writeTo(BufferedSink bufferedSink) {
+        if (progressListener == null) {
+            writeWithoutProgressListener(bufferedSink);
+        } else {
+            writeWithProgressListener(bufferedSink);
+        }
     }
 
+    /**
+     * Returns content length. If the content length cannot be derermined
+     * (is coming from a stream) this value will be -1.
+     * @return Long representing content length
+     */
     @Override
-    public void writeTo(BufferedSink bufferedSink) throws IOException {
-        byte[] buffer = new byte[AbstractBoxMultipartRequest.BUFFER_SIZE];
-        int n = this.inputStream.read(buffer);
-        int totalWritten = 0;
-        while (n != -1) {
-            bufferedSink.write(buffer, 0, n);
-            totalWritten += n;
-            if (progressListener != null) {
-                progressListener.onProgressChanged(totalWritten, this.contentLength());
+    public long contentLength() {
+        return fileSize;
+    }
+
+    private void writeWithoutProgressListener(BufferedSink bufferedSink) {
+        try (Source source = Okio.source(this.inputStream)) {
+            bufferedSink.writeAll(source);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void writeWithProgressListener(BufferedSink bufferedSink) {
+        try {
+            byte[] buffer = new byte[AbstractBoxMultipartRequest.BUFFER_SIZE];
+            int n = this.inputStream.read(buffer);
+            int totalWritten = 0;
+            while (n != -1) {
+                bufferedSink.write(buffer, 0, n);
+                totalWritten += n;
+                if (progressListener != null) {
+                    progressListener.onProgressChanged(totalWritten, this.contentLength());
```

Comment:
What if `contentLength()` return -1. Would progress listener work correctly ?

Reply:
I've added details here: https://github.com/box/box-java-sdk/blob/24b17565570d032f93752036a9d1ae6cfb8ce8f4/src/main/java/com/box/sdk/ProgressListener.java#L10

in short total size will be -1 but bytes written should be properly returned.